internal/cluster/cluster.go

```diff
@@ -177,6 +177,14 @@ func (m *Manager) RegisterAddHostsOCPCluster(c *common.Cluster, db *gorm.DB) err
 }
 
 func (m *Manager) DeregisterCluster(ctx context.Context, c *common.Cluster) error {
+
+	for _, h := range c.Hosts {
+		if err := m.hostAPI.ReportValidationFailedMetrics(ctx, h); err != nil {
+			m.log.WithError(err).Errorf("Failed to report metrics for failed validations on host %s in cluster %s", h.ID, c.ID)
+			return err
```

Comment:
Won't that kill the loop for the first host that failed to report metrics?
Say a cluster consists of 5 hosts, and in the third iteration, it returned `err`, we will never report for the last two.

Perhaps log how many hosts failed to report metrics instead?

Reply:
This is a good point.

We can return the error only at the end of the loop.
@nmagnezi

Reply:
you can count and log something like:
`Failed to report failed validations metrics for 2 out of 5 hosts in cluster <id>`

Reply:
> Although, if we failed, that means that we have a bug and the cluster won't be deleted anyway - do we really want to deliver metrics in case of a bugged service?

That's actually another issue with this return statement: do you want to prevent cluster deletion if we failed to report metrics (whatever the reason may be?)

this logic is activated as a part of Deregister cluster, IMO, it should not stop its execution. I do wonder what others will say.

Reply:
1. You have an option to collect multi error https://github.com/hashicorp/go-multierror
2. Why would you fail the API on metrics?

Reply:
@avishayt what do you think? should we fail delete api if we can't send the metrics?

Reply:
No, we shouldn't fail APIs because of metrics.  It will bring the entire system to a halt.  Should be best effort.

Reply:
How we will know that the "picture" described by the metrics isn't corrupted if we didn't failed on it? we will probably miss the logs. @avishayt 

I mean that if in each cluster the second and third host fail to report metrics you will see only 33% of the `ntpValidation` for example and and may think ntp issues aren't frequent.

Reply:
I was thinking that it would fail here due to reliance on an external metric service, but that's not how it works.  It will fail due to DB error or bad validation format.  So if the DB isn't working we should fail here - the remainder of the deregister probably won't work either.  But if we failed to unmarshal then we will probably always fail and then never delete, so we should ignore the error.  WDYT?

Reply:
I have changed the inner function so at this point we already fetched the data from DB, therefore the only place we can fail on is `Unmarshal`.

If we fail to unmarshal that means that:
1. we will always fail and not deregister
2. bad validation format in DB

Now my question is why is this a problem that the cluster won't deregister?
I mean if we have a bug, lets fail on it, discover the bug and fix it, old cluster are deleted after some time anyway.

IMO this is more important then just cleaning the cluster and never now we have a bug.

If you disagree I will go with skipping the error but if it was only my decision I won't have do that.

What is your conclusion?

Reply:
OK, we can leave as is

Reply:
I will add multi error to catch all hosts at once and resolve the thread.
Thanks :)